rfcs/text/0008_pulse.md

```diff
@@ -0,0 +1,247 @@
+
+# Detailed design
+
+This design is going to be tackled by introducing some common concepts to be used by the main two main components in this architecture:
+
+1. Remote Pulse Service
+2. Local Pulse Service
+
+After that, it explains how we invision the architecture and design of each of those components.
+
+## Concepts
+
+There are some new concepts we'd like to introduce with this new way of reporting telemetry:
+
+- **Deployment Hash ID**  
+This is the _anonymised_ random ID assigned for a deployment. It is used to link multiple pieces of information for further analysis like cross-referencing different bits of information from different sources.
+- **Channels**  
+This is each stream of data that have common information. Typically each channel will have a well defined source of information, different to the rest. It will also result in a different structure to the rest of channels. However, all the channels will maintain a minimum piece of common schema for cross-references (like **Deployment Hash ID** and **timestamp**).
+- **Instructions**  
+These are the messages generated in the form of feedback to the different channels.  
+Typically, channels will follow a bi-directional communication process _(Local <-> Remote)_ but there might be channels that do not generate any kind of instruction _(Local -> Remote)_ and, similarly, some other channels that do not provide any telemetry at all, but allows Pulse to send updates to our products _(Local <- Remote)_.
+
+## Architecture
+
+As mentioned earlier, at the beginning of this chapter, there are two main components in this architecture:
+
+1. Remote Pulse Service
+2. Local Pulse Service
+
+### 1. Remote Pulse Service
+
+This is the service that will receive and store the telemetry from all the _opted-in_ deployments. It will also generate the messages we want to report back to each deployment (aka: instructions).
+
+#### Deployment
+
+- The service will be hosted by Elastic.
+- Most likely maintained by the Infra team.
+- GCP is contemplated at this moment, but we need to confirm how would it affect us regarding the FedRamp approvals (and similar).
+- Exposes an API (check [Endpoints](#endpoints) to know more) to inject the data and retrieve the _instructions_.
+- The data will be stored in an ES cluster.
+
+#### Endpoints
+
+The following endpoints will send every payload detailed in below encrypted with a similar mechanism to the current telemetry encryption.
+
+##### Authenticate
+
+This Endpoint will be used to retrieve a randomised `deploymentID` and a `token` for the cluster to use in all the subsequent requests. Ideally, it will provide some sort of identifier (like `cluster_uuid` or `license.uuid`) so we can revoke its access to any of the endpoints if explicitly requested ([Blocking telemetry input](https://github.com/elastic/telemetry/pull/221) and [Delete previous telemetry data](https://github.com/elastic/telemetry/issues/209)).
```

Comment:
After giving this some more thought and discussing with @epixa, when a user's browser is mediating the communication between the Kibana server and the remote pulse service we'll need additional protections than those already provided by [request-crypto](https://github.com/elastic/request-crypto). `request-crypto` provides the guarantee that the data being sent from the Kibana server can only be decrypted by the Pulse service. However, it does not provide any guarantees to the Kibana server that a response comes from the Pulse service. 

Before addressing these deficiencies, we'll only be able to rely upon the authenticate endpoint when the Kibana server can contact the Pulse service directly.

Reply:
I think we can merge this conversation with https://github.com/elastic/kibana/pull/57108#discussion_r379388005 ?